source/slang/slang-ast-reflect.h

```diff
@@ -28,4 +28,26 @@
 #define SLANG_REFLECTED
 #define SLANG_UNREFLECTED
 
+// Macros for simulating virtual methods without virtual methods
+
+#define SLANG_AST_NODE_INVOKE(method, methodParams) _##method##Override methodParams
+
+#define SLANG_AST_NODE_CASE(NAME, SUPER, ORIGIN, LAST, MARKER, TYPE, param)      case ASTNodeType::NAME: return static_cast<NAME*>(this)-> SLANG_AST_NODE_INVOKE param;
+
+#define SLANG_AST_NODE_VIRTUAL_CALL(base, methodName, methodParams) \
+    switch (astNodeType) \
+    { \
+        SLANG_ALL_ASTNode_##base(SLANG_AST_NODE_CASE, (methodName, methodParams)) \
+        default: return SLANG_AST_NODE_INVOKE (methodName, methodParams); \
+    }
+
+// Same but for a method that's const
+#define SLANG_AST_NODE_CONST_CASE(NAME, SUPER, ORIGIN, LAST, MARKER, TYPE, param)      case ASTNodeType::NAME: return static_cast<const NAME*>(this)-> SLANG_AST_NODE_INVOKE param;
```

Comment:
Minor design nit: it would eliminate some duplication in this code to not bother having a separate `const` and non-`const` cases. My general philosophy is that `const`-ness is mostly only of interest when it pertains to "value" types, and of almost no benefit when it comes to "reference" / "object" types (which is what the AST nodes are).

I know there are differences of opinion on this stuff, so I'm fine with future proofing things for `const` use cases, but I'm also no excited about a future where those distinctions start to matter for stuff like the AST.